File: 2176-count-equal-and-divisible-pairs-in-an-array/2176-count-equal-and-divisible-pairs-in-an-array.go
func countPairs(nums []int, k int) int {
    hashMap := make(map[int][]int)
    ans := 0
    for i, v := range nums {
        if _, ok := hashMap[v]; ok {
            hashMap[v] = append(hashMap[v], i)
        } else {
            hashMap[v] = []int{i}
        }
        if i % k == 0 {
            ans = ans + len(hashMap[v]) - 1
        } else {
            for _, value := range hashMap[v] {
                if value == i {
                    continue
                }
                if value*i % k == 0 {
                    ans++
                }
            }
        }
    }
    return ans
}

File: 2176-count-equal-and-divisible-pairs-in-an-array/2176-count-equal-and-divisible-pairs-in-an-array.ts
function countPairs(nums: number[], k: number): number {
    const hashMap: Map<number, number[]> = new Map<number, number[]>()
    var ans: number = 0
    
    nums.forEach((value:number, index:number) => {
        if (hashMap.has(value)) {
            hashMap.get(value).push(index)
            hashMap.set(value, hashMap.get(value))
        } else {
            hashMap.set(value, [index])
        }
        if (index % k == 0) {
            ans += hashMap.get(value).length - 1
        } else {
            var indexes: number[] = hashMap.get(value)
            indexes.forEach((v:number) => {
                if (v!=index && v*index % k == 0) {
                    ans++
                }
            })
        }
    })
    return ans
};